### PR TITLE
Tell ai what to do #2013

### LINKED ISSUE #2013: [FR] Tell AI what to do next
### Description

Enable users to tell AI what to do next based on previous rounds of AI calls

### Impact

Users who write an article with the help of OpenAI Writer

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by rizwan3395**
### Feature Preview
https://github.com/AppFlowy-IO/AppFlowy/assets/47111784/167bbd12-4eea-4c1e-924d-51ec537917ab

---

#### PR Checklist
Add the option to rewrite AI generated content, make text longer, continue writing, or edit the prompt.

fixes #2013 

- [x] My code adheres to the [AppFlowy Style Guide](https://appflowy.gitbook.io/docs/essential-documentation/contribute-to-appflowy/software-contributions/submitting-code/style-guides)
- [x] I've listed at least one issue that this PR fixes in the description above.
- [x] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [x] All existing tests are passing.

### CODE REVIEW

**Comment by LucasXu0** [comment]
Hey, @rizwan3395. I believe that you should also update the integration test.
```log
❌ integration tests for open-ai smart menu testing selection on open-ai smart menu insert (failed)
  [🌎 Easy Localization] [DEBUG] Localization initialized
  [🌎 Easy Localization] [DEBUG] Start
  ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  │ #0   Log.info (package:appflowy_backend/log.dart:31:26)
  │ #1   NetworkListener._updateConnectionStatus (package:appflowy/core/network_monitor.dart:52:9)
  ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
  │ 💡 Network type: Unknown
  └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  ══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
  The following TestFailure was thrown running a test:
  Expected: Selection:<start = path = [2], offset = 0, end = path = [3], offset = 0>
    Actual: Selection:<start = path = [1], offset = 0, end = path = [1], offset = 80>
  
  When the exception was thrown, this was the stack:
  #4      main.<anonymous closure>.<anonymous closure> (file:///home/runner/work/AppFlowy/AppFlowy/frontend/appflowy_flutter/integration_test/open_ai_smart_menu_test.dart:75:7)
  <asynchronous suspension>
  <asynchronous suspension>
  (elided one frame from package:stack_trace)
  
  This was caught by the test expectation on the following line:
    file:///home/runner/work/AppFlowy/AppFlowy/frontend/appflowy_flutter/integration_test/open_ai_smart_menu_test.dart line 75
  The test description was:
    testing selection on open-ai smart menu insert
```

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart [hunk 1]
```diff
@@ -253,6 +314,8 @@ class _AutoCompletionInputState extends State<_AutoCompletionInput> {
     loading.start();
     await _updateEditingText();
     final result = await UserBackendService.getCurrentUserProfile();
+    int? writeStartPath;
+    int? writeEndPath;
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart:318]: use `Path` instead.

```
Path? start;
Path? end;
```

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart [hunk 2]
```diff
@@ -263,9 +326,12 @@ class _AutoCompletionInputState extends State<_AutoCompletionInput> {
       await openAIRepository.getStreamedCompletions(
         prompt: controller.text,
         onStart: () async {
+          writeStartPath =
+              widget.editorState.document.root.children.last.path.last;
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart:330]: I suppose you can replace it with `node.path.previous`.

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart [hunk 3]
```diff
@@ -279,15 +345,20 @@ class _AutoCompletionInputState extends State<_AutoCompletionInput> {
           }
         },
         onEnd: () async {
+          writeEndPath =
+              widget.editorState.document.root.children.last.path.last;
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart:349]: I suppose you can replace it with `node.path.previous`.

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart [hunk 4]
```diff
@@ -279,15 +345,20 @@ class _AutoCompletionInputState extends State<_AutoCompletionInput> {
...
           widget.editorState.service.selectionService.currentSelection
               .addListener(_onCancelWhenSelectionChanged);
+          textFieldFocusNode.unfocus();
         },
         onError: (error) async {
           loading.stop();
           await _showError(error.message);
         },
       );
+      _updateSelection(writeStartPath, writeEndPath);
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart:360]: await?

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart [hunk 5]
```diff
@@ -297,25 +368,79 @@ class _AutoCompletionInputState extends State<_AutoCompletionInput> {
     });
   }
 
-  Future<void> _onRewrite() async {
-    String previousOutput = _getPreviousOutput()!;
+  Future<void> _updateSelection(
+    start,
+    end,
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart:373]: Add the type for the parameters.

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart [hunk 6]
```diff
@@ -297,25 +368,79 @@ class _AutoCompletionInputState extends State<_AutoCompletionInput> {
...
+        textNodesInselection
+            .add(widget.editorState.document.nodeAtPath([i]) as TextNode);
+      } catch (e) {
+        continue;
+      }
+    }
+    final previousOutput =
+        widget.editorState.getTextInSelection(textNodesInselection, selection);
+    return previousOutput.join(" ");
+  }
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart:421]:
```suggestion
  String _getAllPreviousText() {
    Selection selection = Selection.fromJson(
      jsonDecode(widget.node.attributes[kAutoCompletionSelectionRange]),
    );
    final startPath = selection.start.path;
    final endPath = selection.end.path;
    if (startPath.isEmpty || endPath.isEmpty) {
      return '';
    }
    List<TextNode> textNodes = [];
    for (int i = startPath.last; i < endPath.last; i++) {
      final path = startPath
        ..removeLast()
        ..add(i);
      final node = widget.editorState.document.nodeAtPath(path);
      if (node is TextNode) {
        textNodes.add(node);
      }
    }
    final previousOutput = widget.editorState.getTextInSelection(
      textNodes,
      selection,
    );
    return previousOutput.join(' ');
  }
```

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart [hunk 7]
```diff
@@ -324,8 +449,11 @@ class _AutoCompletionInputState extends State<_AutoCompletionInput> {
         apiKey: userProfile.openaiKey,
       );
       await openAIRepository.getStreamedCompletions(
-        prompt: _rewritePrompt(previousOutput),
+        prompt: action
+            .prompt(action == RewriteAction.editPrompt ? text : previousOutput),
         onStart: () async {
+          writeStartPath =
+              widget.editorState.document.root.children.last.path.last;
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart:456]: same as the previous comment.

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_plugins.dart [hunk 8]
```diff
@@ -10,7 +10,7 @@ SelectionMenuItem autoGeneratorMenuItem = SelectionMenuItem.node(
   iconData: Icons.generating_tokens,
   keywords: ['ai', 'openai' 'writer', 'autogenerator'],
   nodeBuilder: (editorState) {
-    final node = Node(
+    final node = Node(
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_plugins.dart:13]: remove the redundant space.

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart [hunk 9]
```diff
@@ -339,7 +467,12 @@ class _AutoCompletionInputState extends State<_AutoCompletionInput> {
             );
           }
         },
-        onEnd: () async {},
+        onEnd: () async {
+          writeEndPath =
+              widget.editorState.document.root.children.last.path.last;
+          _updateSelection(writeStartPath, writeEndPath);
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart:473]: await ?

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart (hunk 2 above)
Comment (rizwan3395) [reply frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart:330]: In this case the correct path can be obtained only using `widget.editorState.document.root.children.last.path`

For example:

There is a text already written, let's say "Hello world" and the user start using AI after that text.

in that case `widget.node.previous.path` shows the path [1], while `widget.editorState.document.root.children.last.path.last` shows the path [2] and it's working with the path [2].

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart (hunk 3 above)
Comment (rizwan3395) [reply frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart:349]: In this case the correct path can be obtained only using widget.editorState.document.root.children.last.path

For example:

There is a text already written, let's say "Hello world" and the user start using AI after that text.

in that case widget.node.previous.path shows the path [1], while widget.editorState.document.root.children.last.path.last shows the path [2] and it's working with the path [2].

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart [hunk 10]
```diff
@@ -143,12 +143,44 @@ class _AutoCompletionInputState extends State<_AutoCompletionInput> {
...
+                      hoverColor: AFThemeExtension.of(context).lightGreyHover,
+                      width: 22,
+                    ),
+                  ),
+                ],
+              ),
+              const Space(0, 10),
+            ],
+          )
+        : Container();
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart:181]:
```suggestion
        : const SizedBox.shrink();
```

File: frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart [hunk 11]
```diff
@@ -143,12 +143,44 @@ class _AutoCompletionInputState extends State<_AutoCompletionInput> {
...
+                        "editor/close",
+                        color: Theme.of(context).colorScheme.tertiary,
+                      ),
+                      hoverColor: AFThemeExtension.of(context).lightGreyHover,
+                      width: 22,
+                    ),
+                  ),
+                ],
+              ),
+              const Space(0, 10),
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/plugins/openai/widgets/auto_completion_node_widget.dart:178]:
```suggestion
              const VSpace(10),
```

**Comment by LucasXu0** [comment]
I will fix the integrate test later.

**Comment by annieappflowy** [comment]
When can this be released?

**Pull request closed without merging**